platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/IsolatedExtensionsContainer.kt

```diff
@@ -0,0 +1,128 @@
+import org.gradle.internal.extensions.stdlib.uncheckedCast
+import org.gradle.internal.isolate.graph.IsolatedActionDeserializer
+import org.gradle.internal.isolate.graph.IsolatedActionSerializer
+import org.gradle.internal.isolate.graph.SerializedIsolatedActionGraph
+
+internal class IsolatedExtensionsContainer(
+    private val serializer: IsolatedActionSerializer,
+    private val deserializer: IsolatedActionDeserializer,
+    private val delegate: ExtensionContainerInternal,
+    private val gradleProperties : GradleProperties
+) : ExtensionContainerInternal {
+
+    private val serializedExtensionsByType = mutableMapOf<TypeOf<*>, SerializedIsolatedActionGraph<Any>>()
+    private val serializedExtensionsByName = mutableMapOf<String, SerializedIsolatedActionGraph<Any>>()
+    private val serializedExtensionsAsMap: SerializedIsolatedActionGraph<Map<String, Any>> by lazy {
+        serializer.serialize(delegate.getAsMap())
+    }
+    private val serializedExtraProperties : SerializedIsolatedActionGraph<Map<String, Any>> by lazy {
+        serializer.serialize(delegate.extraProperties.properties)
+    }
+
+    override fun getAsMap(): Map<String, Any> =
+        deserializer.deserialize(serializedExtensionsAsMap)
+
+    override fun getExtensionsSchema(): ExtensionsSchema =
+        DefaultExtensionsSchema.create(delegate.extensionsSchema)
+
+    override fun <T : Any> getByType(type: Class<T>): T = getByType(typeOf(type))
+
+    override fun <T : Any> getByType(type: TypeOf<T>): T {
+        val serialized = serializedExtensionsByType.computeIfAbsent(type) {
+            serializer.serialize(delegate.getByType(type))
+        }
+        return deserializer.deserialize(serialized).uncheckedCast()
+    }
+
+    override fun <T : Any> findByType(type: Class<T>): T? = findByType(typeOf(type))
+
+    override fun <T : Any> findByType(type: TypeOf<T>): T? {
+        val serialized = serializedExtensionsByType.compute(type) { _, value ->
+            value ?: delegate.findByType(type)?.let { serializer.serialize(it) }
+        }
+        return serialized?.let { deserializer.deserialize(it).uncheckedCast() }
+    }
+
+    override fun getByName(name: String): Any {
+        val serialized = serializedExtensionsByName.computeIfAbsent(name) {
+            serializer.serialize((delegate.getByName(name)))
+        }
+        return deserializer.deserialize(serialized)
+    }
+
+    override fun findByName(name: String): Any? {
+        val serialized = serializedExtensionsByName.compute(name) { _, value ->
+            value ?: delegate.findByName(name)?.let { serializer.serialize(it) }
+        }
+        return serialized?.let { deserializer.deserialize(it) }
+    }
+
+    override fun getExtraProperties(): ExtraPropertiesExtension {
+        val properties = deserializer.deserialize(serializedExtraProperties)
+        return DefaultExtraPropertiesExtension(properties).apply { setGradleProperties(gradleProperties) }
```

Comment:
I think we should probably use an immutable version of `ExtraPropertiesExtension` here